hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"fmt"
+	"go/token"
+
+	"github.com/dave/dst"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astbuilder"
+)
+
+// ConversionParameters are parameters for converting between a source type and a destination type.
+type ConversionParameters struct {
+	Source            dst.Expr
+	Destination       dst.Expr
+	SourceType        Type
+	DestinationType   Type
+	NameHint          string
+	ConversionContext []Type
+	AssignmentHandler func(destination, source dst.Expr) dst.Stmt
+}
+
+// ClonedSource clones the Source field.
+func (params ConversionParameters) ClonedSource() dst.Expr {
```

Comment:
Do you need to call out that it is cloned? It's kind of an implementation detail.

What about making the field `source` and providing a method `Source()` that just happens to return a clone.

Reply:
Currently this is in the astmodel package and the ARM stuff is in a subpackage, so I can't make these fields private (not exported?) as that would prevent the usage from constructing them.

I could change the methods to `GetSource` + `GetDestination`?

Reply:
What if you introduced a constructor/factory method so that direct construction wasn't needed?

If that's not feasible, then `GetSource()` etc would be fine.

Reply:
The issue with constructor/factory method is that this type has 8 properties, some of which are regularly `nil`. An 8 parameter function is significantly less understandable/readable than an 8 parameter struct instantiation I think.

One approach that could get these names nicer would be to duplicate the `ConversionParameters` structure and turn it into a `parameters` struct that's only used to supply arguments to the factory function. Then make everything not exported on `ConversionParameters` so that we could use `Source` and `Destination`. That's quite a bit of extra code though and feels like it might be a bit overengineered.

Will just change to `GetX` for now I think